File: ios/Library/Source/Auth/SignUpScreenlet/SignUpScreenlet.swift
/**
* Copyright (c) 2000-present Liferay, Inc. All rights reserved.
*
* This library is free software; you can redistribute it and/or modify it under
* the terms of the GNU Lesser General Public License as published by the Free
* Software Foundation; either version 2.1 of the License, or (at your option)
* any later version.
*
* This library is distributed in the hope that it will be useful, but WITHOUT
* ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
* FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
* details.
*/
import UIKit


@objc public protocol SignUpScreenletDelegate {

	optional func onSignUpResponse(attributes: [String:AnyObject])
	optional func onSignUpError(error: NSError)

}


@IBDesignable public class SignUpScreenlet: BaseScreenlet, AnonymousAuthData {

	//MARK: Inspectables

	@IBInspectable public var anonymousApiUserName: String? = "test@liferay.com"
	@IBInspectable public var anonymousApiPassword: String? = "test"

	@IBInspectable public var autologin: Bool = true

	@IBInspectable public var saveCredentials: Bool = true

	@IBInspectable public var companyId: Int64 = 0 {
		didSet {
			(screenletView as? SignUpData)?.companyId = self.companyId
		}
	}

	@IBOutlet public var delegate: SignUpScreenletDelegate?
	@IBOutlet public var autoLoginDelegate: LoginScreenletDelegate?


	internal var signUpData: SignUpData {
		return screenletView as SignUpData
	}


	//MARK: BaseScreenlet

	override func onCreated() {
		super.onCreated()

		signUpData.companyId = self.companyId
	}

	override internal func onUserAction(actionName: String?, sender: AnyObject?) {
		let signUpOperation = LiferaySignUpOperation(screenlet: self)

		signUpOperation.validateAndEnqueue() {
			if $0.lastError != nil {
				self.delegate?.onSignUpError?($0.lastError!)
			}
			else {
				self.onSignUpSuccess(userAttributes: signUpOperation.resultUserAttributes!)
			}
		}
	}


	//MARK: Private methods

	private func onSignUpSuccess(#userAttributes: [String:AnyObject]) {
		delegate?.onSignUpResponse?(userAttributes)

		if autologin {
			SessionContext.removeStoredSession()

			SessionContext.createSession(
					username: signUpData.emailAddress!,
					password: signUpData.password!,
					userAttributes: userAttributes)

			autoLoginDelegate?.onLoginResponse?(userAttributes)

			if saveCredentials {
				if SessionContext.storeSession() {
					autoLoginDelegate?.onCredentialsSaved?()
				}
			}
		}
	}

}

File: android/viewsets/westeros/src/main/java/com/liferay/mobile/screens/viewsets/westeros/auth/signup/SignUpScreenlet.java
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 * <p/>
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 * <p/>
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

package com.liferay.mobile.screens.viewsets.westeros.auth.signup;

import android.content.Context;
import android.util.AttributeSet;
import com.liferay.mobile.screens.auth.signup.interactor.SignUpInteractor;

/**
 * @author Silvio Santos
 */
public class SignUpScreenlet extends com.liferay.mobile.screens.auth.signup.SignUpScreenlet
    implements com.liferay.mobile.screens.viewsets.westeros.auth.signup.SignUpListener {

    public static final String TERMS_AND_CONDITIONS = "TERMS_AND_CONDITIONS";
    private SignUpListener listener;

    public SignUpScreenlet(Context context) {
        super(context);
    }

    public SignUpScreenlet(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public SignUpScreenlet(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public SignUpScreenlet(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    public SignUpListener getListener() {
        return listener;
    }

    public void setListener(SignUpListener value) {
        listener = value;
    }

    @Override
    public void onClickOnTermsAndConditions() {
        if (listener != null) {
            listener.onClickOnTermsAndConditions();
        }
    }

    @Override
    protected void onUserAction(String userActionName, SignUpInteractor interactor, Object... args) {
        if (TERMS_AND_CONDITIONS.equals(userActionName)) {
            onClickOnTermsAndConditions();
        } else {
            super.onUserAction(userActionName, interactor, args);
        }
    }
}